### PR TITLE
add inputmanager module

### PR DESCRIPTION
**Pull request by BurningTreeC**
this stores informations about inputs using a unique id

id
selectionStart
selectionEnd
shouldFocusAgain (input had focus and $:/state/current-focus has the input-id in its text field)

### CODE REVIEW

File: core/modules/inputmanager.js [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+\*/
+(function(){
+
+/*jslint node: true, browser: true */
+/*global $tw: false */
+"use strict";
+
+function InputManager(options) {
+	options = options || "";
```
Comment (pmario) [inline core/modules/inputmanager.js]: Is options a string or an object? IMO it needs to be `|| {}`

File: core/modules/inputmanager.js [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+
+Input handling utilities
+
+\*/
+(function(){
+
+/*jslint node: true, browser: true */
+/*global $tw: false */
+"use strict";
+
```
Comment (pmario) [inline core/modules/inputmanager.js]: Comment block with general info is missing.

File: core/modules/inputmanager.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+function InputManager(options) {
+	options = options || "";
+	this.inputs = [];
+}
+
+InputManager.prototype.findInputById = function(identifier) {
+	var result;
+	for(var i=0; i<this.inputs.length; i++) {
+		var object = this.inputs[i];
```
Comment (pmario) [inline core/modules/inputmanager.js]: why is it named `object` ... is there a better name? .... Is it an html `element` ?

File: core/modules/inputmanager.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+		return false;
+	}
+};
+
+InputManager.prototype.setValue = function(identifier,name,value) {
+	var object = this.findInputById(identifier);
+	if(object) {
+		object[name] = value;
+	} else {
+		var options = {
```
Comment (pmario) [inline core/modules/inputmanager.js]: `options` ... I'm confused. IMO this is different to the first `options` from `function InputManager(options) {` .. right?

File: core/modules/inputmanager.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+			result = object;
+			break;
+		}
+	}
+	return result;
+};
+
+InputManager.prototype.shouldFocusAgain = function(identifier) {
+	var object = this.findInputById(identifier);
+	if(object && object.shouldFocusAgain && $tw.wiki.getTiddlerText("$:/state/current-focus") === identifier) {
```
Comment (pmario) [inline core/modules/inputmanager.js]: hmm. `$tw.wiki` is the global wiki instance. ... imo it should be like this: `this.wiki.getTiddlerText(STATE_CURRENT_FOCUS)`

File: core/modules/inputmanager.js (hunk 1 above)
Comment (pmario) [reply core/modules/inputmanager.js]: I'd add: 

```
this.wiki = options.wiki || $tw.wiki;
```

see comment below!

File: core/modules/inputmanager.js (hunk 2 above)
Comment (pmario) [reply core/modules/inputmanager.js]: IMO there should be a global variable, that contains the tiddler name. It makes it easier to find typos.

```
var STATE_CURRENT_FOCUS = "$:/state/current-focus"
```

it's used in comments below.

File: core/modules/inputmanager.js [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+		var options = {
+			id: identifier
+		};
+		options[name] = value;
+		this.inputs.push(options);
+	}
+};
+
+InputManager.prototype.getSelections = function(identifier) {
+	var object = this.findInputById(identifier);
```
Comment (pmario) [inline core/modules/inputmanager.js]: `object` ... for me that's a confusing var name.

File: core/modules/inputmanager.js [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+		var selEnd = object.selectionEnd;
+		if(selStart && selEnd) {
+			return {
+				selectionStart: selStart,
+				selectionEnd: selEnd
+			};
+		} else {
+			return null;
+		}
+	}
```
Comment (pmario) [inline core/modules/inputmanager.js]: what if the `object` doesn't exist. ... IMO else path is missing.

File: core/modules/inputmanager.js [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+		this.inputs.push(options);
+	}
+};
+
+InputManager.prototype.getSelections = function(identifier) {
+	var object = this.findInputById(identifier);
+	if(object) {
+		var selStart = object.selectionStart;
+		var selEnd = object.selectionEnd;
+		if(selStart && selEnd) {
```
Comment (pmario) [inline core/modules/inputmanager.js]: Are the above 3 lines necessary?

File: core/modules/inputmanager.js (hunk 1 above)
Comment (BurningTreeC) [reply core/modules/inputmanager.js]: I just copied this over from our keyboardManager

File: core/modules/inputmanager.js (hunk 3 above)
Comment (BurningTreeC) [reply core/modules/inputmanager.js]: no it's just an object containing 
```
{
id: uniqueID,
selectionStart: value,
selectionEnd: value,
shoudFocusAgain: boolean
}
```

I think inputInfo is a good name

File: core/modules/inputmanager.js (hunk 4 above)
Comment (BurningTreeC) [reply core/modules/inputmanager.js]: correct. I understand the need to rename variables in order for them to be more self-explanatory

File: core/modules/inputmanager.js (hunk 8 above)
Comment (BurningTreeC) [reply core/modules/inputmanager.js]: no not necessarily

**Comment by BurningTreeC** [comment]
Using this I already have inputs working that edit the current Tiddler

**Comment by pmario** [comment]
So the only thing missing is a little bit more docs that goes with the code.

**Comment by pmario** [comment]
Especially what return values are expected. eg: undefined or null ... or what ever is expected in success case or error case.

**Comment by pmario** [comment]
Nice!

**Comment by BurningTreeC** [comment]
reopening because of #3747

**Comment by BurningTreeC** [comment]
I'm inclined to close this in favor of a new, cleaner PR

**Comment by BurningTreeC** [comment]
closing in favor of #4113

**Pull request closed without merging**